File: internal/trace/state.go
package trace

import (
	"fmt"
	"sort"
)

// State is the algorithm state at some point in a trace, plus the call stack.
//
// The semantics of applying an event live here rather than in a player,
// because "what does this event mean" is a property of the FORMAT. A player is
// navigation on top of these two functions.
//
// STORAGE MODEL, and the JS twin must match it exactly or conformance fails:
//
//   - scalar / array / grid / map : one flat map keyed by at.Join().
//     A missing key reads as Fill.
//   - nodes / graph               : at[0] selects a top-level slot (a node id,
//     "$refs" or "$edges"); further segments descend into Records.
//
// Grid and array are stored SPARSELY (only written cells are present) and read
// through to Fill. That is deliberate: it makes forward-then-backward return to
// a byte-identical state with no special casing, because StateHash skips any
// value equal to Fill. Present-with-fill and absent hash identically.
type State struct {
	Structs map[string]*Struct
	order   []string // struct creation order, for stable non-hash iteration

	// Stack holds the event indices of currently-open calls, oldest first.
	Stack []int

	// retToCall pairs every ret event with its call, computed once in NewState.
	// Backward stepping through a ret needs to know which frame to restore, and
	// scanning for it would make Prev O(n).
	retToCall map[int]int
}

// Struct is one named data container.
type Struct struct {
	Name   string
	Kind   Kind
	Dims   []int
	Fill   Value
	Aux    bool
	Labels *Labels
	Schema *Schema

	flat map[string]Value // scalar | array | grid | map
	root map[string]Value // nodes | graph
	// order is node creation order. Renderers use it for deterministic
	// tie-breaking; nothing depends on it being meaningful.
	order []string
}

// NewState returns an empty state with call pairing precomputed.
func NewState(t *Trace) *State {
	s := &State{
		Structs:   map[string]*Struct{},
		retToCall: map[int]int{},
	}
	var stack []int
	for i, e := range t.Events {
		switch e.T {
		case Call:
			stack = append(stack, i)
		case Ret:
			if n := len(stack); n > 0 {
				s.retToCall[i] = stack[n-1]
				stack = stack[:n-1]
			} else {
				s.retToCall[i] = -1
			}
		}
	}
	return s
}

// Names returns struct names in creation order.
func (s *State) Names() []string { return s.order }

// Get reads through to Fill for absent addresses.
func (s *State) Get(name string, at Path) Value {
	st := s.Structs[name]
	if st == nil {
		return nil
	}
	return st.Get(at)
}

func (st *Struct) Get(at Path) Value {
	if st.Kind.IsNodeKind() {
		return st.getNested(at)
	}
	if v, ok := st.flat[at.Join()]; ok {
		return v
	}
	return st.Fill
}

func (st *Struct) getNested(at Path) Value {
	if len(at) == 0 {
		return st.Fill
	}
	cur, ok := st.root[at[0].String()]
	if !ok {
		return st.Fill
	}
	for _, seg := range at[1:] {
		switch c := cur.(type) {
		case Record:
			v, ok := c[seg.String()]
			if !ok {
				return st.Fill
			}
			cur = v
		case Tuple:
			if !seg.IsIdx || seg.I < 0 || seg.I >= len(c) {
				return st.Fill
			}
			cur = c[seg.I]
		default:
			return st.Fill
		}
	}
	return cur
}

// Set writes v at at. The value is normalized and cloned, so state never
// aliases an event's payload -- without the clone, unapplying a node creation
// would mutate the event that created it.
func (st *Struct) Set(at Path, v Value) {
	v = Clone(Normalize(v))
	if !st.Kind.IsNodeKind() {
		st.flat[at.Join()] = v
		return
	}
	if len(at) == 0 {
		return
	}
	head := at[0].String()
	if len(at) == 1 {
		if _, seen := st.root[head]; !seen {
			st.order = append(st.order, head)
		}
		st.root[head] = v
		return
	}
	// Descend, creating Records on demand. A missing intermediate is an
	// algorithm bug that V4 will have already flagged; creating it here keeps
	// the player total rather than panicking mid-demo.
	container, ok := st.root[head]
	if !ok || container == nil {
		container = Record{}
		if _, seen := st.root[head]; !seen {
			st.order = append(st.order, head)
		}
		st.root[head] = container
	}
	for i := 1; i < len(at)-1; i++ {
		seg := at[i].String()
		rec, ok := container.(Record)
		if !ok {
			rec = Record{}
			st.assign(at[:i], rec)
			container = rec
		}
		next, ok := rec[seg]
		if !ok || next == nil {
			next = Record{}
			rec[seg] = next
		}
		container = next
	}
	if rec, ok := container.(Record); ok {
		rec[at[len(at)-1].String()] = v
	}
}

func (st *Struct) assign(at Path, v Value) {
	if len(at) == 1 {
		st.root[at[0].String()] = v
	}
}

// Exists reports whether a node slot currently holds a non-nil value.
func (st *Struct) Exists(id string) bool {
	v, ok := st.root[id]
	return ok && v != nil
}

// NodeIDs returns node slots in creation order, excluding reserved namespaces.
func (st *Struct) NodeIDs() []string {
	out := make([]string, 0, len(st.order))
	for _, id := range st.order {
		if id == NSRefs || id == NSEdges {
			continue
		}
		out = append(out, id)
	}
	return out
}

// Flat exposes the flat cell map for renderers and tests. Read only.
func (st *Struct) Flat() map[string]Value { return st.flat }

// Root exposes the node map for renderers and tests. Read only.
func (st *Struct) Root() map[string]Value { return st.root }

// ---------------------------------------------------------------------------
// Applying events
// ---------------------------------------------------------------------------

// ApplyForward applies one event. idx is the event's index in Trace.Events,
// needed so a call can be identified by position (calls carry no id).
//
// changed, when non-nil, receives the address keys this event touched. It is
// mutated in place; nothing here allocates a new set.
func (s *State) ApplyForward(idx int, e *Event, changed map[string]bool) error {
	switch e.T {
	case Init:
		if _, dup := s.Structs[e.S]; dup {
			return fmt.Errorf("init: structure %q already exists", e.S)
		}
		st := &Struct{
			Name: e.S, Kind: e.Kind, Dims: e.Dims, Fill: Normalize(e.Fill),
			Aux: e.Aux, Labels: e.Labels, Schema: e.Schema,
		}
		if e.Kind.IsNodeKind() {
			st.root = map[string]Value{}
		} else {
			st.flat = map[string]Value{}
		}
		s.Structs[e.S] = st
		s.order = append(s.order, e.S)
		if changed != nil {
			changed[e.S+" "] = true
		}
	case Set:
		st := s.Structs[e.S]
		if st == nil {
			return fmt.Errorf("set: unknown structure %q", e.S)
		}
		st.Set(e.At, e.To)
		if changed != nil {
			changed[e.At.KeyWith(e.S)] = true
		}
	case Call:
		s.Stack = append(s.Stack, idx)
	case Ret:
		if n := len(s.Stack); n > 0 {
			s.Stack = s.Stack[:n-1]
		}
	}
	return nil
}

// ApplyBackward undoes one event using only that event (invariant I1).
func (s *State) ApplyBackward(idx int, e *Event, changed map[string]bool) error {
	switch e.T {
	case Init:
		delete(s.Structs, e.S)
		if n := len(s.order); n > 0 && s.order[n-1] == e.S {
			s.order = s.order[:n-1]
		} else {
			for i, n := range s.order {
				if n == e.S {
					s.order = append(s.order[:i], s.order[i+1:]...)
					break
				}
			}
		}
		if changed != nil {
			changed[e.S+" "] = true
		}
	case Set:
		st := s.Structs[e.S]
		if st == nil {
			return fmt.Errorf("set: unknown structure %q", e.S)
		}
		st.Set(e.At, e.From)
		if changed != nil {
			changed[e.At.KeyWith(e.S)] = true
		}
	case Call:
		if n := len(s.Stack); n > 0 {
			s.Stack = s.Stack[:n-1]
		}
	case Ret:
		if c, ok := s.retToCall[idx]; ok && c >= 0 {
			s.Stack = append(s.Stack, c)
		}
	}
	return nil
}

// ---------------------------------------------------------------------------
// Hashing
// ---------------------------------------------------------------------------

// Addresses returns every live address key with its canonical value, sorted.
// Values equal to the structure's Fill are SKIPPED -- that is what makes sparse
// and dense storage hash identically, and what makes a rewind land back on the
// initial hash without any special casing.
func (s *State) Addresses() []string {
	names := make([]string, 0, len(s.Structs))
	for n := range s.Structs {
		names = append(names, n)
	}
	sort.Strings(names)

	var out []string
	for _, n := range names {
		st := s.Structs[n]
		if st.Kind.IsNodeKind() {
			keys := make([]string, 0, len(st.root))
			for k := range st.root {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			for _, k := range keys {
				out = appendLeaves(out, st, n+" "+k, st.root[k])
			}
		} else {
			keys := make([]string, 0, len(st.flat))
			for k := range st.flat {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			for _, k := range keys {
				v := st.flat[k]
				if Equal(v, st.Fill) {
					continue
				}
				out = append(out, n+" "+k+"="+Canon(v))
			}
		}
	}
	return out
}

func appendLeaves(out []string, st *Struct, prefix string, v Value) []string {
	if Equal(v, st.Fill) {
		return out
	}
	switch t := v.(type) {
	case Record:
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			out = appendLeaves(out, st, prefix+"/"+k, t[k])
		}
		return out
	}
	return append(out, prefix+"="+Canon(v))
}

// StateHash is FNV-1a over the sorted address list plus the call stack depth.
//
// The JS implementation must produce the same 64-bit value for the same state.
// scripts/conformance.sh diffs the two, step by step, on every golden trace.
func (s *State) Hash() uint64 {
	const (
		offset64 = 14695981039346656037
		prime64  = 1099511628211
	)
	h := uint64(offset64)
	write := func(str string) {
		for i := 0; i < len(str); i++ {
			h ^= uint64(str[i])
			h *= prime64
		}
		h ^= 0
		h *= prime64
	}
	for _, a := range s.Addresses() {
		write(a)
	}
	write("#stack")
	for _, f := range s.Stack {
		write(fmt.Sprintf("%d", f))
	}
	return h
}

File: web/src/player/state.js
// @ts-check
/**
 * The state model, and the semantics of applying an event.
 *
 * This is a deliberate mirror of internal/trace/state.go. The two are written
 * twice ON PURPOSE: two independent implementations that must agree at every
 * step is the mechanism by which the format gets tested, not an accident of
 * having two languages. scripts/conformance.sh diffs their hashes.
 *
 * STORAGE MODEL -- must match the Go twin exactly:
 *
 *   scalar / array / grid / map : one flat Map keyed by at.join('/').
 *                                 A missing key reads as `fill`.
 *   nodes / graph               : at[0] selects a top-level slot (a node id,
 *                                 '$refs' or '$edges'); further segments
 *                                 descend into plain objects.
 *
 * Grid and array are stored SPARSELY and read through to `fill`. That is what
 * makes forward-then-backward land on a byte-identical state with no special
 * casing, because the hash skips any value equal to `fill`: present-with-fill
 * and absent hash the same.
 */

import { addrKey, canon, clone, deepEqual } from '../lib/value.js';

const NS_REFS = '$refs';
const NS_EDGES = '$edges';

const NODE_KINDS = new Set(['nodes', 'graph']);

export class Struct {
  /** @param {object} init the init event */
  constructor(init) {
    this.name = init.s;
    this.kind = init.kind;
    this.dims = init.dims ?? null;
    this.fill = init.fill ?? null;
    this.aux = !!init.aux;
    this.labels = init.labels ?? null;
    this.schema = init.schema ?? null;
    this.isNodeKind = NODE_KINDS.has(this.kind);

    /** @type {Map<string,*>} flat cells, for scalar/array/grid/map */
    this.flat = this.isNodeKind ? null : new Map();
    /** @type {Object<string,*>|null} top-level slots, for nodes/graph */
    this.root = this.isNodeKind ? Object.create(null) : null;
    /** @type {string[]} node creation order */
    this.order = [];
  }

  /** @param {Array<number|string>} at */
  get(at) {
    if (!this.isNodeKind) {
      const k = at.join('/');
      return this.flat.has(k) ? this.flat.get(k) : this.fill;
    }
    if (at.length === 0) return this.fill;
    let cur = this.root[String(at[0])];
    if (cur === undefined) return this.fill;
    for (let i = 1; i < at.length; i++) {
      if (cur === null || typeof cur !== 'object') return this.fill;
      const seg = at[i];
      const next = Array.isArray(cur) ? cur[Number(seg)] : cur[String(seg)];
      if (next === undefined) return this.fill;
      cur = next;
    }
    return cur;
  }

  /**
   * Writes v at at. The value is CLONED: without that, unapplying a node
   * creation would mutate the very event that created it, and the second
   * rewind through it would be wrong.
   * @param {Array<number|string>} at
   */
  set(at, v) {
    const val = clone(v);
    if (!this.isNodeKind) {
      this.flat.set(at.join('/'), val);
      return;
    }
    if (at.length === 0) return;
    const head = String(at[0]);
    if (!(head in this.root)) this.order.push(head);
    if (at.length === 1) {
      this.root[head] = val;
      return;
    }
    let container = this.root[head];
    if (container === undefined || container === null || typeof container !== 'object') {
      container = {};
      this.root[head] = container;
    }
    for (let i = 1; i < at.length - 1; i++) {
      const seg = String(at[i]);
      let next = container[seg];
      if (next === undefined || next === null || typeof next !== 'object') {
        next = {};
        container[seg] = next;
      }
      container = next;
    }
    container[String(at[at.length - 1])] = val;
  }

  /** @param {string} id */
  exists(id) {
    return this.isNodeKind && this.root[id] !== undefined && this.root[id] !== null;
  }

  /** Node slots in creation order, excluding the reserved namespaces. */
  nodeIDs() {
    return this.order.filter((id) => id !== NS_REFS && id !== NS_EDGES);
  }

  /** Named pointers as {name: nodeId}. */
  refs() {
    const out = {};
    const r = this.isNodeKind ? this.root[NS_REFS] : null;
    if (r && typeof r === 'object') {
      for (const k of Object.keys(r)) {
        const v = r[k];
        out[k] = v && typeof v === 'object' && typeof v.$ === 'string' ? v.$ : null;
      }
    }
    return out;
  }

  /** Edge attribute records as {"u|v": {w: ...}}. */
  edges() {
    const e = this.isNodeKind ? this.root[NS_EDGES] : null;
    return e && typeof e === 'object' ? e : {};
  }

  /** @param {Array<number|string>} at */
  key(at) {
    return addrKey(this.name, at);
  }
}

export class State {
  /** @param {object} trace a validated trace */
  constructor(trace) {
    /** @type {Map<string, Struct>} */
    this.structs = new Map();
    /** @type {string[]} creation order */
    this.order = [];
    /** @type {number[]} event indices of open calls, oldest first */
    this.stack = [];

    // Pair every ret with its call, once. Backward stepping through a ret needs
    // to know which frame to restore, and scanning for it would make prev()
    // O(n) instead of O(1).
    /** @type {Map<number, number>} */
    this.retToCall = new Map();
    const st = [];
    const evs = trace.events;
    for (let i = 0; i < evs.length; i++) {
      if (evs[i].t === 'call') st.push(i);
      else if (evs[i].t === 'ret') this.retToCall.set(i, st.length ? st.pop() : -1);
    }
  }

  /** @param {string} name @param {Array<number|string>} at */
  get(name, at) {
    const s = this.structs.get(name);
    return s ? s.get(at) : null;
  }

  /**
   * Apply one event forward.
   *
   * MUST NOT ALLOCATE on the `set` path. Everything about the perceived speed
   * of this app is downstream of that: seeking is O(distance), so one scrubber
   * drag applies thousands of events, and an allocation per apply turns that
   * into GC pressure and a stuttering scrubber. `changed` is mutated in place
   * for the same reason.
   *
   * @param {number} idx event index
   * @param {object} e
   * @param {Set<string>|null} changed mutated in place
   */
  forward(idx, e, changed) {
    switch (e.t) {
      case 'init': {
        const s = new Struct(e);
        this.structs.set(e.s, s);
        this.order.push(e.s);
        if (changed) changed.add(e.s + ' ');
        break;
      }
      case 'set': {
        const s = this.structs.get(e.s);
        if (!s) return;
        s.set(e.at ?? [], e.to);
        if (changed) changed.add(addrKey(e.s, e.at ?? []));
        break;
      }
      case 'call':
        this.stack.push(idx);
        break;
      case 'ret':
        this.stack.pop();
        break;
    }
  }

  /** Undo one event using only that event (invariant I1). */
  backward(idx, e, changed) {
    switch (e.t) {
      case 'init': {
        this.structs.delete(e.s);
        const i = this.order.lastIndexOf(e.s);
        if (i >= 0) this.order.splice(i, 1);
        if (changed) changed.add(e.s + ' ');
        break;
      }
      case 'set': {
        const s = this.structs.get(e.s);
        if (!s) return;
        s.set(e.at ?? [], e.from);
        if (changed) changed.add(addrKey(e.s, e.at ?? []));
        break;
      }
      case 'call':
        this.stack.pop();
        break;
      case 'ret': {
        const c = this.retToCall.get(idx);
        if (c !== undefined && c >= 0) this.stack.push(c);
        break;
      }
    }
  }

  /**
   * Every live address with its canonical value, sorted.
   *
   * Values equal to the structure's `fill` are SKIPPED. That is what makes
   * sparse and dense storage hash identically, and what makes a rewind land
   * back on the initial hash with no special casing.
   * @returns {string[]}
   */
  addresses() {
    const names = [...this.structs.keys()].sort();
    const out = [];
    for (const n of names) {
      const s = this.structs.get(n);
      if (s.isNodeKind) {
        for (const k of Object.keys(s.root).sort()) {
          appendLeaves(out, s, n + ' ' + k, s.root[k]);
        }
      } else {
        for (const k of [...s.flat.keys()].sort()) {
          const v = s.flat.get(k);
          if (deepEqual(v, s.fill)) continue;
          out.push(n + ' ' + k + '=' + canon(v));
        }
      }
    }
    return out;
  }

  /**
   * FNV-1a over the sorted address list plus the call stack.
   * Must produce the same 64-bit value as trace.State.Hash in Go.
   * @returns {bigint}
   */
  hash() {
    const MASK = (1n << 64n) - 1n;
    const PRIME = 1099511628211n;
    let h = 14695981039346656037n;
    const write = (str) => {
      for (let i = 0; i < str.length; i++) {
        h = (h ^ BigInt(str.charCodeAt(i) & 0xff)) & MASK;
        h = (h * PRIME) & MASK;
      }
      h = (h * PRIME) & MASK; // the trailing 0x00 separator
    };
    for (const a of this.addresses()) write(a);
    write('#stack');
    for (const f of this.stack) write(String(f));
    return h;
  }
}

function appendLeaves(out, s, prefix, v) {
  if (deepEqual(v, s.fill)) return;
  if (v !== null && typeof v === 'object' && !Array.isArray(v) && typeof v.$ !== 'string') {
    for (const k of Object.keys(v).sort()) appendLeaves(out, s, prefix + '/' + k, v[k]);
    return;
  }
  out.push(prefix + '=' + canon(v));
}
